Support DML operations on Delta Lake tables with `id` column mapping

findinpath opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

This PR is a follow-up of https://github.com/trinodb/trino/pull/16183 and relates to https://github.com/trinodb/trino/issues/12638

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues



https://books.japila.pl/delta-lake-internals/DeltaConfigs/#COLUMN_MAPPING_MODE

id | A column ID is the identifier of a column. This mode is used for tables converted from Iceberg and parquet files in this mode will also have corresponding field Ids for each column in their file schema.
-- | --

I intended to test this column mapping mode for tables migrated from Iceberg and used

https://docs.databricks.com/sql/language-manual/delta-convert-to-delta.html


Apply `CLONE` on an Iceberg table created by Trino was causing internal error issues on Databricks, so I used Spark with Iceberg to create the table on AWS and then applied `CLONE` via Databricks.

```
CREATE OR REPLACE TABLE default.findinpath_table_migrated CLONE iceberg.`s3://trino-bucket/iceberg-warehouse/findinpath.db/findinpath_table/metadata/00001-0b3c1ecd-e5b9-479f-9ddd-e286b9efd472.metadata.json`
``` 

Outcome was a table with table writer version 7 , which can't be written by Trino. :( 

Nevertheless, relevant info - the Parquet files written via Iceberg look like:

```
➜  ~ parquet-tools schema  /Users/marius/Downloads/00001-3-c0e1f76e-1234-4de8-9348-b18789e0720f-00001.parquet 
message table {
  required binary c (STRING) = 1;
}
```

The new Parquet schema for the files written by Databricks (on `INSERT` statements after the `CLONE` statement) look like 

```
➜  ~ parquet-tools schema /Users/marius/Downloads/part-00000-cdd011ba-6032-4012-9cb7-364ca60a5b1f-c000.snappy\ \(1\).parquet
message spark_schema {
  optional binary col-34630ee4-4dbf-4916-a464-5f13f9d3ff46 (STRING) = 1;
}
```

The parquet files produced by Trino with the column mapping mode set to `id` are also consistent with this schema.

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Support DML operations on Delta Lake tables with `id` column mapping ({issue}`issuenumber`)
```

findepi commented:
/test-with-secrets sha=4e6dae7c622f3fffa809c8cce7fff6c587334505

raunaqmorarka commented on plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeMetadataSchemas.java:
```diff
@@ -0,0 +1,349 @@
+            OptionalInt id,
+            DeltaLakeSchemaSupport.ColumnMappingMode columnMappingMode,
+            List<String> parent,
+            BiConsumer<List<String>, Type> primitiveTypesConsumer)
```
Why not pass in the builder instead of `BiConsumer` ? Not sure why we would need a more generic parameter here.

raunaqmorarka commented on plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeMetadataSchemas.java:
```diff
@@ -0,0 +1,349 @@
+            trinoType = typeManager.fromSqlType(primitiveType);
+            verify(trinoType instanceof DecimalType, "type %s does not map to Trino decimal".formatted(primitiveType));
+            DecimalType trinoDecimalType = (DecimalType) trinoType;
+            typeBuilder = Types.primitive(PrimitiveType.PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY, repetition)
```
We should use INT32 for precision <= 9, INT64 for short decimals, then fixed length byte array for all other cases.

raunaqmorarka commented on plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeMetadataSchemas.java:
```diff
@@ -0,0 +1,349 @@
+                typeBuilder = Types.primitive(PrimitiveType.PrimitiveTypeName.INT64, repetition).as(LogicalTypeAnnotation.timestampType(false, LogicalTypeAnnotation.TimeUnit.MILLIS));
+                trinoType = TIMESTAMP_MILLIS;
+            }
+            default -> throw new TypeNotFoundException(new TypeSignature(primitiveType));
```
This should probably be `throw new TrinoException(NOT_SUPPORTED, format("Unsupported primitive type: %s", primitiveType))` because the type is not unknown to us, just that we don't support it yet.

raunaqmorarka commented on plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeMetadataSchemas.java:
```diff
@@ -0,0 +1,349 @@
+        Types.PrimitiveBuilder<PrimitiveType> typeBuilder;
+        Type trinoType;
+        if (primitiveType.startsWith(StandardTypes.DECIMAL)) {
+            trinoType = typeManager.fromSqlType(primitiveType);
```
If this string is something coming out of delta, then should be using Trino's types parser ? If this is something specified in delta standard then maybe we can write code to parse it here.

ebyhr commented on plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:
```diff
@@ -1466,9 +1465,10 @@ public void finishMerge(ConnectorSession session, ConnectorMergeTableHandle tabl
-                    : handle.getMetadataEntry().getOriginalPartitionColumns();
+            List<String> partitionColumns = switch (getColumnMappingMode(handle.getMetadataEntry())) {
+                case NAME, ID -> getPartitionColumnsForNameOrIdMapping(handle.getMetadataEntry().getOriginalPartitionColumns(), ((DeltaLakeMergeTableHandle) tableHandle).getInsertTableHandle().getInputColumns());
+                default -> handle.getMetadataEntry().getOriginalPartitionColumns();
```
No, we should throw an exception for `UNKNOWN`.

findinpath commented:
@ebyhr  could you please run this PR with secrets?

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:
```diff
@@ -134,6 +141,131 @@ public void testColumnMappingMode(String mode)
+                " LOCATION 's3://" + bucketName + "/databricks-compatibility-test-" + tableName + "'" +
+                " TBLPROPERTIES (" +
+                " 'delta.columnMapping.mode'='" + mode + "'," +
+                " 'delta.minReaderVersion'='2'," +
+                " 'delta.minWriterVersion'='5')");
```
nit: Can we remove these versions from the table definition?

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:
```diff
@@ -134,6 +141,131 @@ public void testColumnMappingMode(String mode)
+                "    a_map MAP<STRING, STRING>," +
+                "    a_complex_map MAP<STRING, STRUCT<a_string: STRING>>," +
+                "    a_struct STRUCT<a_string: STRING, a_int: INT>," +
+                "    a_complex_struct STRUCT<nested_struct: STRUCT<a_string: STRING>, a_int: INT>" +
```
Specifying all types in **_a_** table reduces the readability. Is testing with data provider per type is too slow?

findinpath replied:
Do note that we have 3 column mappings x 2 (partitioned/unpartitioned) already.
I opted intentionally in this case to trade test runtime for readability.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:
```diff
@@ -840,21 +962,7 @@ public void testUnsupportedColumnMappingModeChangeDataFeed(String mode)
-                        .hasMessageContaining("Unsupported column mapping mode for tables with change data feed enabled: id");
-            }
-            else if (mode.equals("name")) {
+            if (mode.equals("name") || mode.equals("id")) {
```
This condition looks redundant.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:
```diff
@@ -134,6 +141,131 @@ public void testColumnMappingMode(String mode)
+                    row(true, 1, 10, 100, 1000L, new BigDecimal("123.12"), new BigDecimal("123456789012345678.123"), 0d, 0f, "a", java.sql.Date.valueOf(LocalDate.of(2020, 8, 21)), new byte[] {(byte) 0xAB, (byte) 0xCD}, "element 1", "nested 1", "value1", "nested value1", "item 1", 1, "nested item 1", 11),
+                    row(true, 2, 20, 200, 2000L, new BigDecimal("223.12"), new BigDecimal("223456789012345678.123"), 0d, 0f, "b", java.sql.Date.valueOf(LocalDate.of(2020, 8, 22)), new byte[] {(byte) 0xAB, (byte) 0xCD}, "element 2", "nested 2", null, null, "item 2", 2, "nested item 2", 22));
+
+            String selectDatabricksValues = "SELECT " +
```
nit: I would rename to `selectDeltaValues` for compatibility with `onDelta` because it's not specific to Databricks.

findinpath commented on plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeMetadataSchemas.java:
```diff
@@ -0,0 +1,358 @@
+import static org.apache.parquet.schema.Type.Repetition.OPTIONAL;
+import static org.apache.parquet.schema.Type.Repetition.REQUIRED;
+
+public final class DeltaLakeMetadataSchemas
```
Renamed to `DeltaLakeParquetSchemas`

findinpath commented on plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeMetadataSchemas.java:
```diff
@@ -0,0 +1,358 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.deltalake.transactionlog;
```
Moved to the package `io.trino.plugin.deltalake`

findinpath commented on plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeMetadataSchemas.java:
```diff
@@ -0,0 +1,358 @@
+            }
+            case "timestamp" -> {
+                // Spark/DeltaLake stores timestamps in UTC, but renders them in session time zone.
+                // For more info, see https://delta-users.slack.com/archives/GKTUWT03T/p1585760533005400
```
I've taken it from here
https://github.com/trinodb/trino/blob/de4cbc39c07d3efc6e0dd78b76e45d39f25390f0/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java#L515-L517

Should I remove it from there as well?

findinpath commented:
Regarding https://github.com/trinodb/trino/pull/16600#issuecomment-1514340264 , I created the method `io.trino.plugin.deltalake.DeltaLakeCdfPageSink#adapt` to add the `_change_type` column to the Parquet schema  of CDF files.

It feels to me like a patch method, but I didn't find a better alternative that doesn't involve more refactorings.
Another option would have been to have a similar hierarchy of `TypeParquetMetadata` (`PrimitiveTypeParquetMetadata`, `RowTypeParquetMetadata`, `ArrayTypeParquetMetadata`, `MapTypeParquetMetadata`) to get from `DeltaLakeSchemaSupport` into `DeltaLakeColumnMetadata` and `DeltaLakeColumnHandle` so that it can be passed to `ParquetSchemaConverter` and used for making use of `id`, `physicalName`.

findinpath commented:
Given that  adding `_change_type` column is a rather exceptional use-case, I've decided to add handling for it in the method 
`io.trino.plugin.deltalake.DeltaLakeParquetSchemas#createParquetSchemaMapping(io.trino.plugin.deltalake.transactionlog.MetadataEntry, io.trino.spi.type.TypeManager, boolean)`.

I think the PR is now ready for review.

pajaks commented on plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:
```diff
@@ -1778,7 +1791,7 @@ private void checkWriteSupported(ConnectorSession session, SchemaTableName schem
         checkUnsupportedGeneratedColumns(metadataEntry);
         ColumnMappingMode columnMappingMode = getColumnMappingMode(metadataEntry);
-        if (!(columnMappingMode == NONE || columnMappingMode == ColumnMappingMode.NAME)) {
+        if (!(columnMappingMode == NONE || columnMappingMode == ColumnMappingMode.NAME || columnMappingMode == ColumnMappingMode.ID)) {
```
Can you add static import for NAME and ID similar to NONE?

pajaks commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:
```diff
@@ -485,22 +615,12 @@ public void testUnsupportedOperationsColumnMappingMode(String mode)
-                assertQueryFailure(() -> onTrino().executeQuery("UPDATE delta.default." + tableName + " SET a_string = 'test'"))
-                        .hasMessageContaining("Writing with column mapping id is not supported");
-            }
+            assertThat(onTrino().executeQuery("INSERT INTO delta.default." + tableName + " VALUES (1, 'one'), (2, 'two')"))
```
As test is named  `testUnsupportedOperationsColumnMappingMode` can we have separate test for write operation and test only unsupported here?

findinpath replied:
Good point. The DML operations here are actually irrelevant now.
I'll remove them.

pajaks commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:
```diff
@@ -840,33 +960,17 @@ public void testUnsupportedColumnMappingModeChangeDataFeed(String mode)
             onDelta().executeQuery("INSERT INTO default." + targetTableName + " VALUES (3, 'nation3', 300)");
 
             // Column mapping mode 'none' is tested in TestDeltaLakeDatabricksChangeDataFeedCompatibility
-            // TODO: Remove these failure check and update TestDeltaLakeDatabricksChangeDataFeedCompatibility when adding support the column mapping mode
```
This TODO still valid right?

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:
```diff
@@ -134,6 +141,129 @@ public void testColumnMappingMode(String mode)
+                    (partitioned ? ", 'part2'" : "") +
+                    ")");
+
+            List<Row> expectedRows = ImmutableList.of(
```
Can we change to something like:
```java
            Row firstRow = row(false, 1, 10, 100, 1000L, new BigDecimal("123.12"), new BigDecimal("123456789012345678.123"), 0d, 0f, "a", java.sql.Date.valueOf(LocalDate.of(2020, 8, 21)), new byte[] {(byte) 0xAB, (byte) 0xCD}, "element 1", "nested 1", "value1", "nested value1", "item 1", 1, "nested item 1", 11);
            Row secondRow = row(true, 2, 20, 200, 2000L, new BigDecimal("223.12"), new BigDecimal("223456789012345678.123"), 0d, 0f, "b", java.sql.Date.valueOf(LocalDate.of(2020, 8, 22)), new byte[] {(byte) 0xAB, (byte) 0xCD}, "element 2", "nested 2", null, null, "item 2", 2, "nested item 2", 22);

            List<Row> expectedRows = ImmutableList.of(firstRow, secondRow);
...
            Row updatedFirstRow = new Row(ImmutableList.builder().add(false).addAll(firstRow.getValues().subList(1, firstRow.getValues().size())).build());
...
```

findinpath reviewed (commented):
AC

findinpath commented:
Rebasing on `master` to address conflicts.

findinpath commented:
Rebased again on `master` to address conflicts.
Changed `io.trino.plugin.deltalake.DeltaLakePageSinkProvider#createPageSink(io.trino.spi.connector.ConnectorTransactionHandle, io.trino.spi.connector.ConnectorSession, io.trino.spi.connector.ConnectorOutputTableHandle, io.trino.spi.connector.ConnectorPageSinkId)` to referenced `DeltaLakeColumnHandle#getColumnName` as consequence of 1b9b9883b92dbfebabae16adb8551f8c2b87b142

findepi commented:
/test-with-secrets sha=1e886c943c9b3f9f4649cb1da2df9addf0f386a3

findinpath commented:
Rebasing on top of `master` (no conflicts met) to ensure that the functionality still behaves as expected with the latest updates.

ebyhr commented:
/test-with-secrets sha=e23f4b626ac58f8a31c07068fc6ba49c774e9a24
